xen: code quality updates and generic patch deduplication

SigmaSquadron opened the pull request:
## Description of changes

- Hopefully increases code quality for pkgs.xen.
- Deduplicates identical patches.
- Moves `generic.nix` and other generic files and patches to their own directory, to not clutter the root directory.

**See individual commits for a more detailed changelog.**

There are many commits in this PR, and they were separated so they could be reviewed more easily. Let me know which ones should be squashed.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
- [x] Tested, as applicable:
  - [x] `pkg-config` test passes.
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review pr 333948"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Effectively reverts #333764 and partially reverts #331867.
---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SigmaSquadron committed:
xen: move generic files to generic/ directory

Most patches used in the Xen build are generic, so let's keep everything
that applies to all versions in one folder.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: deduplicate pkgs.xen and pkgs.xenPackages.xen

There is no point in having both. The top-level package now points
directly to the latest version.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: dehardcode pname

This is useful for the future when we begin building custom versions of
Xen, such as `qubes-vmm-xen`.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: call the postPatch for each pre-fetched source manually instead of going through withTools

withTools and withPrefetchedSources are pretty complicated functions
meant to generalise per-version calls to build phases by each
pre-fetched source. This is step 1 in deprecating them.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: drop systemd dependency from 4.19 and above

From the 4.19 release notes:

When building with Systemd support (./configure --enable-systemd),
remove libsystemd as a build dependency. Systemd Notify support is
retained, now using a standalone library implementation.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: fully remove qemu-traditional

We were still building some minor parts of qemu-traditional by not
disabling it explicitly.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: ensure xsm-policy is being built when FLASK is enabled

Flask is disabled by default, but this will save someone an
overrideAttrs overlay if they're using FLASK.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: deduplicate common patches and drop figs

The IPXE patch is the same across all versions.
Let us put generic patches in the new generic/ directory.

We also disable figs, as they were broken.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: make meta.maintainers more friendly for new contributors

From Peder Sundt:
> I read maintainers = with lib.maintainers; [ ]; as a friendly open
> invitation, while maintainers = [ ]; as a sad state of reality.
> I want people to join the project hence I very much prefer the former.

I don't plan on leaving anytime soon, but let's not make it
more difficult for new maintainers to step up.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: deprecate withPrefetchedSources and withTools functions

We now use easier to understand functions that are properly documented
and aren't as generic.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron commented:
Result of `nixpkgs-review pr 333948` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>libvmi</li>
    <li>qubes-core-vchan-xen</li>
  </ul>
</details>
<details>
  <summary>53 packages built:</summary>
  <ul>
    <li>qemu_xen (qemu_xen_4_19)</li>
    <li>qemu_xen.debug (qemu_xen_4_19.debug)</li>
    <li>qemu_xen.ga (qemu_xen_4_19.ga)</li>
    <li>qemu_xen_4_16</li>
    <li>qemu_xen_4_16.debug</li>
    <li>qemu_xen_4_16.ga</li>
    <li>qemu_xen_4_17</li>
    <li>qemu_xen_4_17.debug</li>
    <li>qemu_xen_4_17.ga</li>
    <li>qemu_xen_4_18</li>
    <li>qemu_xen_4_18.debug</li>
    <li>qemu_xen_4_18.ga</li>
    <li>xen</li>
    <li>xen-guest-agent</li>
    <li>xen-slim</li>
    <li>xen-slim.boot</li>
    <li>xen-slim.dev</li>
    <li>xen-slim.doc</li>
    <li>xen-slim.man</li>
    <li>xen.boot</li>
    <li>xen.dev</li>
    <li>xen.doc</li>
    <li>xen.man</li>
    <li>xenPackages.xen_4_16</li>
    <li>xenPackages.xen_4_16-slim</li>
    <li>xenPackages.xen_4_16-slim.boot</li>
    <li>xenPackages.xen_4_16-slim.dev</li>
    <li>xenPackages.xen_4_16-slim.doc</li>
    <li>xenPackages.xen_4_16-slim.man</li>
    <li>xenPackages.xen_4_16.boot</li>
    <li>xenPackages.xen_4_16.dev</li>
    <li>xenPackages.xen_4_16.doc</li>
    <li>xenPackages.xen_4_16.man</li>
    <li>xenPackages.xen_4_17</li>
    <li>xenPackages.xen_4_17-slim</li>
    <li>xenPackages.xen_4_17-slim.boot</li>
    <li>xenPackages.xen_4_17-slim.dev</li>
    <li>xenPackages.xen_4_17-slim.doc</li>
    <li>xenPackages.xen_4_17-slim.man</li>
    <li>xenPackages.xen_4_17.boot</li>
    <li>xenPackages.xen_4_17.dev</li>
    <li>xenPackages.xen_4_17.doc</li>
    <li>xenPackages.xen_4_17.man</li>
    <li>xenPackages.xen_4_18</li>
    <li>xenPackages.xen_4_18-slim</li>
    <li>xenPackages.xen_4_18-slim.boot</li>
    <li>xenPackages.xen_4_18-slim.dev</li>
    <li>xenPackages.xen_4_18-slim.doc</li>
    <li>xenPackages.xen_4_18-slim.man</li>
    <li>xenPackages.xen_4_18.boot</li>
    <li>xenPackages.xen_4_18.dev</li>
    <li>xenPackages.xen_4_18.doc</li>
    <li>xenPackages.xen_4_18.man</li>
  </ul>
</details>

emilazy commented on pkgs/applications/virtualization/xen/4.18/0000-xen-remove-figs-4.18.patch:
I think it would make sense to drop the removal of the `.fig`/`.eps` files from this patch, as it doesn’t affect the build and bloats the Nixpkgs repository. (I don’t suppose it’s being upstreamed via something like GitHub that we could easily `fetchpatch2`, rather than a mailing list or something?)

SigmaSquadron replied:
So, turns out we didn't need fig2dev at all. I've just dropped the dependency, and though it says "some documentation won't be built", that is mostly a lie.

emilazy reviewed (changes requested):
Haven’t done a proper review yet, just a drive‐by suggestion.

SigmaSquadron committed:
xen: simplify make flags.

XEN_EXTFILES_URL wasn't working back when this override did anything,
and now we bypass it entirely. The LD variable was rewritten to use
lib.meta.getExe.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: prefer flags to pipes or variables in bash scripts

For the update script and the one-liner in the README, use flags that do
what we want to do instead of piping the output to different commands or
using environment variables.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: add more sanitation options to update script

We use a lot of pipes, so it's good to exit if any of the dependent
commands fail.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
xen: avoid writing upstreamPatches multiple times

Use `with upstreamPatches;` instead.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron commented:
Once more, do let me know which commits should be squashed.

SigmaSquadron committed:
xen: only add maintainers if version is security-supported

This prevents users browsing the package in search.nixos.org from
messaging maintainers about an EOL Xen.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

emilazy reviewed (approved):
Great stuff as always! As ever, I don’t have the means to test this myself, but have read over the commits and checked the build. I have left some comments, but most are merely informative and none are at all blocking, so please do ping me in a day or so if I haven’t merged this regardless of any action on or response to them.

I don’t really feel any of these commits would be better squashed. Thank you for making well‐factored PRs that are easy to review commit‐by‐commit.

Result of `nixpkgs-review pr 333948` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>libvmi</li>
    <li>qubes-core-vchan-xen</li>
  </ul>
</details>
<details>
  <summary>53 packages built:</summary>
  <ul>
    <li>qemu_xen (qemu_xen_4_19)</li>
    <li>qemu_xen.debug (qemu_xen_4_19.debug)</li>
    <li>qemu_xen.ga (qemu_xen_4_19.ga)</li>
    <li>qemu_xen_4_16</li>
    <li>qemu_xen_4_16.debug</li>
    <li>qemu_xen_4_16.ga</li>
    <li>qemu_xen_4_17</li>
    <li>qemu_xen_4_17.debug</li>
    <li>qemu_xen_4_17.ga</li>
    <li>qemu_xen_4_18</li>
    <li>qemu_xen_4_18.debug</li>
    <li>qemu_xen_4_18.ga</li>
    <li>xen (xenPackages.xen_4_19)</li>
    <li>xen-guest-agent</li>
    <li>xen-slim (xenPackages.xen_4_19-slim)</li>
    <li>xen-slim.boot (xenPackages.xen_4_19-slim.boot)</li>
    <li>xen-slim.dev (xenPackages.xen_4_19-slim.dev)</li>
    <li>xen-slim.doc (xenPackages.xen_4_19-slim.doc)</li>
    <li>xen-slim.man (xenPackages.xen_4_19-slim.man)</li>
    <li>xen.boot (xenPackages.xen_4_19.boot)</li>
    <li>xen.dev (xenPackages.xen_4_19.dev)</li>
    <li>xen.doc (xenPackages.xen_4_19.doc)</li>
    <li>xen.man (xenPackages.xen_4_19.man)</li>
    <li>xenPackages.xen_4_16</li>
    <li>xenPackages.xen_4_16-slim</li>
    <li>xenPackages.xen_4_16-slim.boot</li>
    <li>xenPackages.xen_4_16-slim.dev</li>
    <li>xenPackages.xen_4_16-slim.doc</li>
    <li>xenPackages.xen_4_16-slim.man</li>
    <li>xenPackages.xen_4_16.boot</li>
    <li>xenPackages.xen_4_16.dev</li>
    <li>xenPackages.xen_4_16.doc</li>
    <li>xenPackages.xen_4_16.man</li>
    <li>xenPackages.xen_4_17</li>
    <li>xenPackages.xen_4_17-slim</li>
    <li>xenPackages.xen_4_17-slim.boot</li>
    <li>xenPackages.xen_4_17-slim.dev</li>
    <li>xenPackages.xen_4_17-slim.doc</li>
    <li>xenPackages.xen_4_17-slim.man</li>
    <li>xenPackages.xen_4_17.boot</li>
    <li>xenPackages.xen_4_17.dev</li>
    <li>xenPackages.xen_4_17.doc</li>
    <li>xenPackages.xen_4_17.man</li>
    <li>xenPackages.xen_4_18</li>
    <li>xenPackages.xen_4_18-slim</li>
    <li>xenPackages.xen_4_18-slim.boot</li>
    <li>xenPackages.xen_4_18-slim.dev</li>
    <li>xenPackages.xen_4_18-slim.doc</li>
    <li>xenPackages.xen_4_18-slim.man</li>
    <li>xenPackages.xen_4_18.boot</li>
    <li>xenPackages.xen_4_18.dev</li>
    <li>xenPackages.xen_4_18.doc</li>
    <li>xenPackages.xen_4_18.man</li>
  </ul>
</details>

SigmaSquadron commented on pkgs/applications/virtualization/xen/generic/default.nix:
```diff
@@ -701,11 +705,14 @@ stdenv.mkDerivation (finalAttrs: {
       mit
     ];
-    maintainers = with lib.maintainers; [ sigmasquadron ];
+    # This automatically removes maintainers from EOL versions of Xen, so we aren't bothered about versions we don't explictly support.
```
4.16 will be EOL in December. Once that happens, the number will be updated to only support 4.17 and above, but the sources for 4.16 will be kept. Once 4.17 becomes EOL, the files relating to 4.16 will be removed, and so on.

One thing I haven't figured out yet is how to let people know the files were removed from Nixpkgs if they try to install a very EOL version. Should I add an entry to every removed Xen version to top-level/aliases.nix, or would entries in packages.nix in the Xen tree be more appropriate?

emilazy replied:
I guess it depends on what package set people are most likely to get their Xen from. `throw`s in package sets can be awkward because of Hydra eval stuff. I think that it is fine that it turns from a big “this is EOL” insecurity warning to just an error without any special `throw` message, though. We often don’t add `throw`s for packages in package sets.

SigmaSquadron replied:
Ah, then we don't need any special setup.

SigmaSquadron commented on pkgs/applications/virtualization/xen/update.sh:
```diff
@@ -1,8 +1,9 @@
 # shellcheck disable=SC2206,SC2207 shell=bash
-set -e
-
+set -o errexit
+set -o pipefail
+set -o nounset
```
Oh yes, but this is a relatively simple script that `errexit` should handle just fine.

Hopefully.

Maybe.

SigmaSquadron commented on pkgs/applications/virtualization/xen/generic/default.nix:
```diff
@@ -664,7 +664,7 @@ stdenv.mkDerivation (finalAttrs: {
       mit
     ];
-    maintainers = [ lib.maintainers.sigmasquadron ];
+    maintainers = with lib.maintainers; [ sigmasquadron ];
```
`with lib.maintainers;` seems preferable in any situation IMO. If there are no maintainers, it's easier for new maintainers to add their name. If there is a single maintainer, it's easier for more maintainers to add their names. If there are multiple maintainers, you avoid writing lib.maintainers multiple times.

SigmaSquadron commented on pkgs/applications/virtualization/xen/generic/default.nix:
```diff
@@ -132,37 +137,93 @@ let
+        # Iterate on each pre-fetched source.
+        builtins.map (
+          source:
+          # Iterate on each path.
```
```suggestion
          # Iterate on each available patch.
```

SigmaSquadron commented:
Gotta love that git rewrites all commits despite the changes only being on like two of those.

Fixed the wording in some code comments, and added more documentation for those updating the Xen files.

emilazy commented:
LGTM; let’s land this, any further improvements can be separate PRs. Cross your fingers for Hydra managing to build everything this time!

SigmaSquadron commented:
[All builds succeeded](https://hydra.nixos.org/dashboard/alpha@sigmasquadron.net#tabs-my-jobs)! Thank you again for your help, emily.

emilazy commented:
Any time :) Really happy to see the Xen stuff come together, feel free to ping me for reviews at your leisure.
